File: Anul 2/Medii de Programare si Proiectare/mpp-proiect-java-Mothyque/backend/Server/src/main/java/org/example/service/Service.java
package org.example.service;

import org.example.domain.Entity;
import org.example.repository.Repository;
import org.example.utils.Observer;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Service<ID, E extends Entity<ID>>
{
    protected final Repository<ID, E> repository;

    private List<Observer> observers = new ArrayList<>();

    public Service(Repository<ID, E> repository)
    {
        this.repository = repository;
    }

    public void add(E entity)
    {
        repository.save(entity);
        updateObservers();
    }

    public void delete(ID id)
    {
        repository.delete(id);
        updateObservers();
    }

    public void update(E entity)
    {
        repository.update(entity);
        updateObservers();
    }

    public Optional<E> findOne(ID id)
    {
        return repository.findOne(id);
    }

    public Iterable<E> findAll()
    {
        return repository.findAll();
    }

    public int size()
    {
        return repository.size();
    }

    public void addObserver(Observer observer)
    {
        observers.add(observer);
    }

    public void updateObservers()
    {
        for (Observer observer : observers)
        {
            observer.update();
        }
    }
}

File: Anul 2/Medii de Programare si Proiectare/Lab/mpp-proiect-csharp-Mothyque/ProiectMPP/ProiectMPP.Server/Service/Service.cs
using ProiectMPP.Common.Domain;
using ProiectMPP.Common.Utils;
using ProiectMPP.Server.Repository;
using System.Collections.Generic;

namespace ProiectMPP.Server.Service
{
    public class Service<TId, TE> where TE : Entity<TId>
    {
        protected readonly IRepository<TId, TE> repository;
        private readonly List<IObserver> observers = new List<IObserver>();

        public Service(IRepository<TId, TE> repository)
        {
            this.repository = repository;
        }

        public virtual void Add(TE entity)
        {
            repository.Save(entity);
            UpdateObservers();
        }

        public virtual void Delete(TId id)
        {
            repository.Delete(id);
            UpdateObservers();
        }

        public virtual void Update(TE entity)
        {
            repository.Update(entity);
            UpdateObservers();
        }

        public virtual TE FindOne(TId id)
        {
            return repository.FindOne(id);
        }

        public virtual IEnumerable<TE> FindAll()
        {
            return repository.FindAll();
        }

        public virtual int Size()
        {
            return repository.Size();
        }

        public void BeginTransaction()
        {
            repository.BeginTransaction();
        }

        public void CommitTransaction()
        {
            repository.CommitTransaction();
        }

        public void RollbackTransaction()
        {
            repository.RollbackTransaction();
        }

        public void AddObserver(IObserver observer)
        {
            observers.Add(observer);
        }

        public void UpdateObservers()
        {
            foreach (var observer in observers)
            {
                observer.Update();
            }
        }
    }
}